File: specs/typespec/org/suborgs.ts
import { type ValidationError, newValidationError } from "../common/common";

const SUBORG_NAME_MAX_LENGTH = 64;

export const ERR_SUBORG_NAME_REQUIRED = "name is required";
export const ERR_SUBORG_NAME_TOO_LONG = "name must be at most 64 characters";
export const ERR_SUBORG_NEW_NAME_REQUIRED = "new_name is required";
export const ERR_SUBORG_NEW_NAME_TOO_LONG =
	"new_name must be at most 64 characters";
export const ERR_SUBORG_REGION_REQUIRED = "pinned_region is required";
export const ERR_SUBORG_EMAIL_REQUIRED = "email_address is required";

// SubOrg is the response type for SubOrg reads.
export interface SubOrg {
	name: string;
	pinned_region: string;
	status: string;
	created_at: string;
}

// SubOrgMember is a member of a SubOrg.
export interface SubOrgMember {
	email_address: string;
	full_name?: string;
	assigned_at: string;
}

// CreateSubOrgRequest is the request body for POST /org/create-suborg.
export interface CreateSubOrgRequest {
	name: string;
	pinned_region: string;
}

export function validateCreateSubOrgRequest(
	request: CreateSubOrgRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.name) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_REQUIRED));
	} else if (request.name.length > SUBORG_NAME_MAX_LENGTH) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_TOO_LONG));
	}

	if (!request.pinned_region) {
		errs.push(newValidationError("pinned_region", ERR_SUBORG_REGION_REQUIRED));
	}

	return errs;
}

// ListSubOrgsRequest is the request body for POST /org/list-suborgs.
export interface ListSubOrgsRequest {
	filter_status?: string;
	pagination_key?: string;
	limit?: number;
}

export function validateListSubOrgsRequest(
	request: ListSubOrgsRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (
		request.filter_status !== undefined &&
		request.filter_status !== "active" &&
		request.filter_status !== "disabled"
	) {
		errs.push(
			newValidationError(
				"filter_status",
				"filter_status must be 'active' or 'disabled'"
			)
		);
	}

	return errs;
}

// ListSubOrgsResponse is the response for POST /org/list-suborgs.
export interface ListSubOrgsResponse {
	suborgs: SubOrg[];
	next_pagination_key: string;
}

// RenameSubOrgRequest is the request body for POST /org/rename-suborg.
export interface RenameSubOrgRequest {
	name: string;
	new_name: string;
}

export function validateRenameSubOrgRequest(
	request: RenameSubOrgRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.name) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_REQUIRED));
	} else if (request.name.length > SUBORG_NAME_MAX_LENGTH) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_TOO_LONG));
	}

	if (!request.new_name) {
		errs.push(newValidationError("new_name", ERR_SUBORG_NEW_NAME_REQUIRED));
	} else if (request.new_name.length > SUBORG_NAME_MAX_LENGTH) {
		errs.push(newValidationError("new_name", ERR_SUBORG_NEW_NAME_TOO_LONG));
	}

	return errs;
}

// DisableSubOrgRequest is the request body for POST /org/disable-suborg.
export interface DisableSubOrgRequest {
	name: string;
}

export function validateDisableSubOrgRequest(
	request: DisableSubOrgRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.name) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_REQUIRED));
	}

	return errs;
}

// EnableSubOrgRequest is the request body for POST /org/enable-suborg.
export interface EnableSubOrgRequest {
	name: string;
}

export function validateEnableSubOrgRequest(
	request: EnableSubOrgRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.name) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_REQUIRED));
	}

	return errs;
}

// AddSubOrgMemberRequest is the request body for POST /org/add-suborg-member.
export interface AddSubOrgMemberRequest {
	name: string;
	email_address: string;
}

export function validateAddSubOrgMemberRequest(
	request: AddSubOrgMemberRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.name) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_REQUIRED));
	}

	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_SUBORG_EMAIL_REQUIRED));
	}

	return errs;
}

// RemoveSubOrgMemberRequest is the request body for POST /org/remove-suborg-member.
export interface RemoveSubOrgMemberRequest {
	name: string;
	email_address: string;
}

export function validateRemoveSubOrgMemberRequest(
	request: RemoveSubOrgMemberRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.name) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_REQUIRED));
	}

	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_SUBORG_EMAIL_REQUIRED));
	}

	return errs;
}

// ListSubOrgMembersRequest is the request body for POST /org/list-suborg-members.
export interface ListSubOrgMembersRequest {
	name: string;
	pagination_key?: string;
}

export function validateListSubOrgMembersRequest(
	request: ListSubOrgMembersRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.name) {
		errs.push(newValidationError("name", ERR_SUBORG_NAME_REQUIRED));
	}

	return errs;
}

// ListSubOrgMembersResponse is the response for POST /org/list-suborg-members.
export interface ListSubOrgMembersResponse {
	members: SubOrgMember[];
	next_pagination_key: string;
}

File: specs/typespec/org/suborgs.go
package org

import (
	"fmt"

	"vetchium-api-server.typespec/common"
)

const (
	subOrgNameMaxLength = 64

	errSubOrgNameRequired    = "name is required"
	errSubOrgNameTooLong     = "name must be at most 64 characters"
	errSubOrgNewNameRequired = "new_name is required"
	errSubOrgNewNameTooLong  = "new_name must be at most 64 characters"
	errSubOrgRegionRequired  = "pinned_region is required"
	errSubOrgEmailRequired   = "email_address is required"
)

// SubOrg is the response type for SubOrg reads.
type SubOrg struct {
	Name         string `json:"name"`
	PinnedRegion string `json:"pinned_region"`
	Status       string `json:"status"`
	CreatedAt    string `json:"created_at"`
}

// SubOrgMember is a member of a SubOrg.
type SubOrgMember struct {
	EmailAddress string  `json:"email_address"`
	FullName     *string `json:"full_name,omitempty"`
	AssignedAt   string  `json:"assigned_at"`
}

// CreateSubOrgRequest is the request body for POST /org/create-suborg.
type CreateSubOrgRequest struct {
	Name         string `json:"name"`
	PinnedRegion string `json:"pinned_region"`
}

func (r CreateSubOrgRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	} else if len(r.Name) > subOrgNameMaxLength {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameTooLong)))
	}

	if r.PinnedRegion == "" {
		errs = append(errs, common.NewValidationError("pinned_region", fmt.Errorf(errSubOrgRegionRequired)))
	}

	return errs
}

// ListSubOrgsRequest is the request body for POST /org/list-suborgs.
type ListSubOrgsRequest struct {
	FilterStatus  *string `json:"filter_status,omitempty"`
	PaginationKey *string `json:"pagination_key,omitempty"`
	Limit         *int    `json:"limit,omitempty"`
}

func (r ListSubOrgsRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.FilterStatus != nil && *r.FilterStatus != "active" && *r.FilterStatus != "disabled" {
		errs = append(errs, common.NewValidationError("filter_status", fmt.Errorf("filter_status must be 'active' or 'disabled'")))
	}

	return errs
}

// ListSubOrgsResponse is the response for POST /org/list-suborgs.
type ListSubOrgsResponse struct {
	SubOrgs           []SubOrg `json:"suborgs"`
	NextPaginationKey string   `json:"next_pagination_key"`
}

// RenameSubOrgRequest is the request body for POST /org/rename-suborg.
type RenameSubOrgRequest struct {
	Name    string `json:"name"`
	NewName string `json:"new_name"`
}

func (r RenameSubOrgRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	} else if len(r.Name) > subOrgNameMaxLength {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameTooLong)))
	}

	if r.NewName == "" {
		errs = append(errs, common.NewValidationError("new_name", fmt.Errorf(errSubOrgNewNameRequired)))
	} else if len(r.NewName) > subOrgNameMaxLength {
		errs = append(errs, common.NewValidationError("new_name", fmt.Errorf(errSubOrgNewNameTooLong)))
	}

	return errs
}

// DisableSubOrgRequest is the request body for POST /org/disable-suborg.
type DisableSubOrgRequest struct {
	Name string `json:"name"`
}

func (r DisableSubOrgRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	}

	return errs
}

// EnableSubOrgRequest is the request body for POST /org/enable-suborg.
type EnableSubOrgRequest struct {
	Name string `json:"name"`
}

func (r EnableSubOrgRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	}

	return errs
}

// AddSubOrgMemberRequest is the request body for POST /org/add-suborg-member.
type AddSubOrgMemberRequest struct {
	Name         string `json:"name"`
	EmailAddress string `json:"email_address"`
}

func (r AddSubOrgMemberRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	}

	if r.EmailAddress == "" {
		errs = append(errs, common.NewValidationError("email_address", fmt.Errorf(errSubOrgEmailRequired)))
	}

	return errs
}

// RemoveSubOrgMemberRequest is the request body for POST /org/remove-suborg-member.
type RemoveSubOrgMemberRequest struct {
	Name         string `json:"name"`
	EmailAddress string `json:"email_address"`
}

func (r RemoveSubOrgMemberRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	}

	if r.EmailAddress == "" {
		errs = append(errs, common.NewValidationError("email_address", fmt.Errorf(errSubOrgEmailRequired)))
	}

	return errs
}

// ListSubOrgMembersRequest is the request body for POST /org/list-suborg-members.
type ListSubOrgMembersRequest struct {
	Name          string  `json:"name"`
	PaginationKey *string `json:"pagination_key,omitempty"`
}

func (r ListSubOrgMembersRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	}

	return errs
}

// ListSubOrgMembersResponse is the response for POST /org/list-suborg-members.
type ListSubOrgMembersResponse struct {
	Members           []SubOrgMember `json:"members"`
	NextPaginationKey string         `json:"next_pagination_key"`
}
